src/librustc_borrowck/borrowck/mod.rs

```diff
@@ -1135,6 +1093,102 @@ impl<'a, 'tcx> BorrowckCtxt<'a, 'tcx> {
         }
     }
 
+    fn note_and_explain_mutbl_error(&self, db: &mut DiagnosticBuilder, err: &BckError<'tcx>,
+                                    error_span: &Span) {
+        match err.cmt.note {
+            mc::NoteClosureEnv(upvar_id) | mc::NoteUpvarRef(upvar_id) => {
+                // If this is an `Fn` closure, it simply can't mutate upvars.
+                // If it's an `FnMut` closure, the original variable was declared immutable.
+                // We need to determine which is the case here.
+                let kind = match err.cmt.upvar().unwrap().cat {
+                    Categorization::Upvar(mc::Upvar { kind, .. }) => kind,
+                    _ => bug!()
+                };
+                if kind == ty::ClosureKind::Fn {
+                    db.span_help(self.tcx.map.span(upvar_id.closure_expr_id),
+                                 "consider changing this closure to take \
+                                 self by mutable reference");
+                }
+            }
+            _ => {
+                let mut cmt = &err.cmt;
+                while let Categorization::Deref(ref inner_cmt, ..) = cmt.cat {
+                    cmt = inner_cmt;
+                }
+                if let Categorization::Local(local_id) = cmt.cat {
+                    let parent = self.tcx.map.get_parent_node(local_id);
+                    let may_be_function = match self.tcx.map.get(parent) {
```

Comment:
I think you can use this funky `FnLikeNode` thing. Something like this instead of that big match:

```rust
use rustc::hir::map::blocks::FnLikeNode;
let opt_function =
    FnLikeNode::from_node(self.tcx.map.get(parent))
    .map(|fn_like| fn_like.decl());
```

I also changed the name of `may_be_function` to `opt_function`, which I find a bit more obvious (since in Rust the type is `Option`, not `Maybe`).